src/desktop/lib/psi/AppProvider.js

```diff
@@ -0,0 +1,34 @@
+import PropTypes from 'prop-types'
+import React, { Component } from 'react'
+
+export class AppProvider extends Component {
+  static childContextTypes = {
+    provide: PropTypes.object,
+  }
+
+  getChildContext() {
+    return {
+      provide: this.props.provide,
+    }
+  }
+
+  render() {
+    return <div>{this.props.children}</div>
+  }
+}
+
+export function injectContext(getContextSlice, WrappedComponent) {
+  return class RelayRouterContextInjector extends Component {
+    static contextTypes = {
+      provide: PropTypes.object.isRequired,
+    }
+
+    render() {
+      const contextSlice = getContextSlice(this.context.provide)
+
+      return (
+        <WrappedComponent {...contextSlice} {...this.props} {...this.state} />
+      )
+    }
+  }
+}
```

Comment:
Can this component be reconciled with the `Artsy` one from `Reaction` somehow? I’d like to contain context as much as possible and now there’s two contexts that hold a Relay environment.